diff --git a/meson.build b/meson.build
index 90d9c0f..6e71531 100644
--- a/meson.build
+++ b/meson.build
@@ -402,6 +402,7 @@ add_project_link_arguments(link_flags, language: ['c', 'cpp', 'objc'])
 if darwin
     path_source = files('osdep/path-darwin.c')
     timer_source = files('osdep/timer-darwin.c')
+    sources += files('osdep/utils-mac.c')
 endif
 
 cocoa = dependency('appleframeworks', modules: ['Cocoa', 'IOKit', 'QuartzCore'],
@@ -415,7 +416,6 @@ if features['cocoa']
     dependencies += cocoa
     path_source += files('osdep/path-mac.m')
     sources += files('osdep/language-mac.c',
-                     'osdep/utils-mac.c',
                      'osdep/mac/app_bridge.m',
                      'player/clipboard/clipboard-mac.m')
     main_fn_source = files('osdep/main-fn-mac.c')
